lib/action.ml

```diff
@@ -11,6 +11,48 @@ let action_error msg = raise (Action_error msg)
 let log = Log.from "action"
 
 module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
+  let username_to_slack_id_tbl = Stringtbl.empty ()
+
+  let canonicalize_username username =
+    username |> String.split_on_chars ~on:[ '.'; '-' ] |> String.concat |> String.lowercase
+
+  let canonicalize_email_username email =
+    email
+    |> Re2.replace_exn ~f:(fun _ -> "") (Re2.create_exn "\\+[^]]*|@[^]]*") (* Remove email domain and everything after '+' *)
+    |> canonicalize_username
+
+  let refresh_username_to_slack_id_tbl ~ctx =
+    match%lwt Slack_api.list_users ~ctx () with
+    | Error e ->
+      log#warn "couldn't fetch list of Slack users: %s" e;
+      Lwt.return_unit
+    | Ok res ->
+      List.iter res.members
+        ~f:(fun user ->
+          match user.profile.email with
+          | None -> ()
+          | Some email ->
+          let username = canonicalize_email_username email in
+          log#info "logging email username %s" username;
+          Stringtbl.update username_to_slack_id_tbl username ~f:(fun _ -> user.id)
+          );
+      Lwt.return_unit
+
+  let match_github_user_to_slack_id ?(force_query = false) ~ctx user =
+    let%lwt () =
+      if force_query
+      then refresh_username_to_slack_id_tbl ~ctx
+      else Lwt.return_unit
+    in
+    let canonicalized_login = canonicalize_username user.login in
+    match Stringtbl.find username_to_slack_id_tbl canonicalized_login with
+    | Some slack_id -> Lwt.return_some slack_id
+    | None ->
+    let%lwt () = refresh_username_to_slack_id_tbl ~ctx in
```

Comment:
Right, the background lwt now only refreshes every 24 hours, I forgot to reflect that in the PR description above. Is this okay? If not, I can also add checking for duration since last update every time we query for a user

Reply:
Indeed, seems good.
On a side note, given you depend on devkit, you could make use of `timely_loop` https://github.com/ahrefs/devkit/blob/master/lwt_util.ml#L11 
It doesn't simplify the code much, so this is more a suggestion than anything.